### PR TITLE
Refactor minor activities

### PR DESCRIPTION
**Pull request by KrisXV**

### CODE REVIEW

File: server/chat-plugins/announcements.ts [hunk 1]
```diff
@@ -2,20 +2,16 @@
  * Announcements chat plugin
  * By Spandamn
  */
-export class Announcement {
-	readonly activityId: 'announcement';
-	announcementNumber: number;
-	room: Room;
+export class Announcement extends Rooms.MinorActivity {
+	activityid = 'announcement' as ID;
```
Comment (AnnikaCodes) [inline server/chat-plugins/announcements.ts:6]: Is there a reason it's all lowercase? My personal preference is for `activityID`; I guess this is more consistent with `roomid` and `userid`, though.

**Review by AnnikaCodes — COMMENTED** [review]
I like this change.

File: server/rooms.ts [hunk 2]
```diff
@@ -861,8 +888,8 @@ export abstract class BasicRoom {
 			connection,
 			'|init|chat\n|title|' + this.title + '\n' + userList + '\n' + this.log.getScrollback() + this.getIntroMessage(user)
 		);
-		if (this.minorActivity) this.minorActivity.onConnect(user, connection);
-		if (this.game && this.game.onConnect) this.game.onConnect(user, connection);
+		if (this.minorActivity?.onConnect) this.minorActivity.onConnect(user, connection);
+		if (this.game?.onConnect) this.game.onConnect(user, connection);
```
Comment (Zarel) [inline server/rooms.ts:892]:
```suggestion
		this.minorActivity?.onConnect?.(user, connection);
		this.game?.onConnect?.(user, connection);
```

If you're gonna optional chain you might as well go all the way.

File: server/rooms.ts [hunk 3]
```diff
@@ -878,8 +905,8 @@ export abstract class BasicRoom {
 		this.users[user.id] = user;
 		this.userCount++;
 
-		if (this.minorActivity) this.minorActivity.onConnect(user, connection);
-		if (this.game && this.game.onJoin) this.game.onJoin(user, connection);
+		if (this.minorActivity?.onConnect) this.minorActivity.onConnect(user, connection);
+		if (this.game?.onJoin) this.game.onJoin(user, connection);
```
Comment (Zarel) [inline server/rooms.ts:909]: (ditto)

File: server/rooms.ts [hunk 4]
```diff
@@ -903,9 +930,7 @@ export abstract class BasicRoom {
 		} else {
 			this.reportJoin('n', user.getIdentityWithStatus(this.roomid) + '|' + oldid, user);
 		}
-		if (this.minorActivity && 'voters' in this.minorActivity) {
-			if (user.id in this.minorActivity.voters) this.minorActivity.updateFor(user);
-		}
+		if (this.minorActivity?.onRename) this.minorActivity.onRename(user, oldid, joining);
```
Comment (Zarel) [inline server/rooms.ts:933]: (ditto)

**Review by Zarel — APPROVED** [review]
I like this! Other than the minor readability adjustment.

**Comment by Zarel** [comment]
(Haha get it "minor"????)

**Comment by mia-pi-git** [comment]
Reminder to add conversion functions for pre-existing roomsettings.

File: server/chat-plugins/announcements.ts [hunk 5]
```diff
@@ -3,27 +3,26 @@
...
+	readonly activityid: 'announcement';
 }
 
-export class Announcement extends MinorActivity {
-	readonly activityId: 'announcement';
-	announcementNumber: number;
+export class Announcement extends Rooms.MinorActivity {
+	readonly activityid = 'announcement' as ID;
+	name = 'Announcement';
+	onRename = undefined;
```
Comment (mia-pi-git) [inline server/chat-plugins/announcements.ts:20]: Why not just make this optional and not put the undefined declare here?

File: server/chat-plugins/announcements.ts [hunk 6]
```diff
@@ -34,44 +33,40 @@ export class Announcement extends MinorActivity {
...
 		this.displayTo(user, connection);
 	}
 
-	end() {
-		this.endTimer();
-		this.room.send(`|uhtmlchange|announcement${this.announcementNumber}|<div class="infobox">(${this.room.tr`The announcement has ended.`})</div>`);
-		delete this.room.settings.minorActivity;
-		this.room.minorActivity = null;
-		this.room.saveSettings();
+	endActivity() {
```
Comment (mia-pi-git) [inline server/chat-plugins/announcements.ts:51]: Rename to `destroy`? More consistent with PS APIs.

File: server/chat-plugins/poll.ts [hunk 7]
```diff
@@ -222,7 +160,7 @@ export class Poll extends MinorActivity {
 		return output;
 	}
 
-	static generateResults(options: PollData, room: Room, ended = false, choice: number[] | null = null) {
+	static generateResults(options: PollData | MinorActivityData, room: Room, ended = false, choice: number[] | null = null) {
```
Comment (mia-pi-git) [inline server/chat-plugins/poll.ts:163]:
```suggestion
	static generateResults(
		options: PollData | MinorActivityData, 
		room: Room, 
		ended = false, 
		choice: number[] | null = null
	) {
```

**Comment by KrisXV** [comment]
> Reminder to add conversion functions for pre-existing roomsettings.

yeah, just trying to fix this crash first then i will

File: server/chat-plugins/poll.ts [hunk 8]
```diff
@@ -398,9 +342,6 @@ export class Poll extends MinorActivity {
 		}
 		return out;
 	}
-	destroy() {
```
Comment (mia-pi-git) [inline server/chat-plugins/poll.ts:401]: Ditto on destroy.

File: server/room-minor-activity.ts [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+	endTimer() {
+		if (!this.timeout) return false;
+		clearTimeout(this.timeout);
+		this.timeoutMins = 0;
+		this.timerEnd = 0;
+		return true;
+	}
+
+	abstract display(): void;
+	abstract endActivity?(): void;
```
Comment (mia-pi-git) [inline server/room-minor-activity.ts:112]: Ditto.

**Review by mia-pi-git — CHANGES_REQUESTED** [review]
Looking good otherwise!
Good change.

File: server/chat-plugins/announcements.ts (hunk 5 above)
Comment (KrisXV) [reply server/chat-plugins/announcements.ts:20]: it _is_ optional, TS is just being buggy

File: server/chat-plugins/announcements.ts (hunk 6 above)
Comment (KrisXV) [reply server/chat-plugins/announcements.ts:51]: sure

File: server/chat-plugins/announcements.ts (hunk 1 above)
Comment (KrisXV) [reply server/chat-plugins/announcements.ts]: i'd prefer activityID too, but i did it for consistency yeah

File: server/room-minor-activity.ts [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+		clearTimeout(this.timeout);
+		this.timeoutMins = 0;
+		this.timerEnd = 0;
+		return true;
+	}
+
+	abstract destroy(): void;
+	abstract display(): void;
+	abstract onConnect?(user: User, connection: Connection | null): void;
+	abstract onRename?(user: User, oldid: ID, joining: boolean): void;
```
[the remark this answers is not in the document]
Comment (KrisXV) [reply server/room-minor-activity.ts:116]: oh ok

File: server/rooms.ts [hunk 11]
```diff
@@ -889,8 +928,8 @@ export abstract class BasicRoom {
 		this.users[user.id] = user;
 		this.userCount++;
 
-		if (this.minorActivity) this.minorActivity.onConnect(user, connection);
-		if (this.game && this.game.onJoin) this.game.onJoin(user, connection);
+		this.minorActivity?.onConnect?.(user, connection);
```
[the remark this answers is not in the document]
Comment (KrisXV) [reply server/rooms.ts:931]: I do not know

File: server/rooms.ts [hunk 12]
```diff
@@ -914,9 +953,7 @@ export abstract class BasicRoom {
 		} else {
 			this.reportJoin('n', user.getIdentityWithStatus(this.roomid) + '|' + oldid, user);
 		}
-		if (this.minorActivity && 'voters' in this.minorActivity) {
-			if (user.id in this.minorActivity.voters) this.minorActivity.updateFor(user);
-		}
+		this.minorActivity?.onRename?.(user, oldid, joining);
```
[the remark this answers is not in the document]
Comment (KrisXV) [reply server/rooms.ts:956]: I do not know

**Review by mia-pi-git — APPROVED** [review]
Except for the onConnect thing, looks good to me!
Good changes.

**Pull request merged**